### PR TITLE
fix(PlayCookie) PLAY_TOKEN cookie rejected because userprofile exceeds 4096 chars

### PR DESCRIPTION
**Pull request by neojunjie**
Refer to the issue created earlier (https://github.com/datahub-project/datahub/issues/4448)
Instead of using play cookie to store the profile returned from pac4j, you are able to use play cache to store these data instead. This will resolve issue where the response returned is greater than 4096 chars and the browser rejects the play_token cookie
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Review by jjoyce0510 — COMMENTED** [review]
Hey @neojunjie ! Thanks for the PR!

The downside to this approach is that a play cache is unique for each node of the datahub-frontend service. Until now, we've attempted to have the datahub-frontend container remain as a stateless service to make it more easily scalable.

Now one possible solution is to modify our recommendations around deploying datahub, particularly on Helm, and recommend that folks use sticky routing to ensure the same user routes to the same frontend node. But that may require some additional changes..

@dexter-mh-lee @RyanHolstien Interested to get your thoughts on this one

John

**Comment by neojunjie** [comment]
Hi ***@***.****

Another way is to use distributed cache? Play do support distributed
caching using memcached and redis?

I think we can take reference from this project  (
https://github.com/mumoshu/play2-memcached)

**Comment by jjoyce0510** [comment]
@neojunjie But that would require yet another dependency on a caching layer. We are hoping to keep the deployment footprint as small as possible to make operating DataHub simpler. 

I think we'll need to have some further consideration for this one - I do like the idea of moving the pac4j profiles outside of the cookie itself since we've seen the issue so frequently

**Comment by RyanHolstien** [comment]
> @neojunjie But that would require yet another dependency on a caching layer. We are hoping to keep the deployment footprint as small as possible to make operating DataHub simpler.
> 
> I think we'll need to have some further consideration for this one - I do like the idea of moving the pac4j profiles outside of the cookie itself since we've seen the issue so frequently

This might make sense as an optional setting for users that are willing to complicate their deployment. I definitely agree that we don't want this to be the default, but making it optional opens up a few different deployment configurations:

1) Default, cookie store -> supports single + multi

2) Single frontend pod, in memory cache

3) Multiple frontend pods, memcached or redis distributed cache

I think the biggest downside to this approach is supporting these use cases and the complicated configurations that come along with them, but we don't currently have a great solution to the cookie being too large so it might be worth supporting.

**Comment by jjoyce0510** [comment]
@neojunjie Yeah you're right - I think having it as something which is easily configurable makes a lot of sense. Then we can make this the default solve for cases where people are exceeding that 4096 character limit. Would you be willing to try to generalize the PR to support easily configuring it via an environment variable?

**Comment by neojunjie** [comment]
@jjoyce0510 Sure! Let me update this PR to make cookie/cache configurable via env variable.

**Comment by jjoyce0510** [comment]
@neojunjie - Checking in here! Have you had time to look into updating this?

**Comment by neojunjie** [comment]
@jjoyce0510  Sorry! Will get it done soon. Was busy with work

**Comment by jjoyce0510** [comment]
Re-running checks against latest. Hoping to get this one in!

File: datahub-frontend/conf/application.conf [hunk 1]
```diff
@@ -39,6 +39,11 @@ play.http.server.akka.max-header-count = ${?DATAHUB_AKKA_MAX_HEADER_COUNT}
 play.server.akka.max-header-value-length = 8k
 play.server.akka.max-header-value-length = ${?DATAHUB_AKKA_MAX_HEADER_VALUE_LENGTH}
 
+# pac4j configuration
+# default to PlayCookieSessionStore to keep datahub-frontend's statelessness
+pac4j.sessionStore.provider= "PlayCookieSessionStore"
+pac4j.sessionStore.provider= ${?PAC4J_SESSIONSTORE_PROVIDER}
```
Comment (jjoyce0510) [inline datahub-frontend/conf/application.conf:45]: Amazing! I'm wondering if we could document this - at least in our FAQ or debugging guide...

Comment (neojunjie) [reply datahub-frontend/conf/application.conf:45]: any idea where I should add this in?

Comment (jjoyce0510) [reply datahub-frontend/conf/application.conf:45]: For now, how about just adding an FAQ here- https://datahubproject.io/docs/debugging 

@RyanHolstien is working on revamping our deployment docs, perhaps we can also add a note in there once they are live. 

Thanks again, @neojunjie !

Comment (neojunjie) [reply datahub-frontend/conf/application.conf:45]: got it. will add them in 
working on the other pr. =(

Comment (neojunjie) [reply datahub-frontend/conf/application.conf:45]: @jjoyce0510 added them in debugging.md. Anything else to add?

Comment (jjoyce0510) [reply datahub-frontend/conf/application.conf:45]: I think that works for now.. Thank you!

**Review by jjoyce0510 — APPROVED** [review]
Thank you @neojunjie !

**Pull request merged**